pkg/bloombuild/planner/queue/queue.go

```diff
@@ -0,0 +1,173 @@
+package queue
+
+import (
+	"context"
+	"fmt"
+	"path/filepath"
+	"sync"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/grafana/dskit/services"
+	"github.com/prometheus/client_golang/prometheus"
+
+	"github.com/grafana/loki/v3/pkg/queue"
+	"github.com/grafana/loki/v3/pkg/util"
+)
+
+type Task interface {
+	Tenant() string
+	Table() string
+	ID() string
+}
+
+// Queue is a wrapper of queue.RequestQueue that uses the file system to store the pending tasks.
+// When a task is enqueued, it's stored in the file system and recorded ad pending.
+// When it's dequeued, it's removed from the queue but kept in FS until removed.
+type Queue[T Task] struct {
```

Comment:
We should not make the mistake of making the package generic. It solves a very specific use-case, and is also located in the `pkg/bloombuild/planner/` package.

Generics make code more difficult to manage than needed

Reply:
Got too fancy there. Removed the generics.